contracts/crashpunks-v2-friedger.clar

```diff
@@ -0,0 +1,345 @@
+(define-constant ERR-NFT-NOT-LISTED-FOR-SALE (err u105))
+(define-constant ERR-PAYMENT-ADDRESS (err u106))
+(define-constant ERR-NFT-LISTED (err u107))
+(define-constant ERR-COLLECTION-LIMIT-REACHED (err u108))
+(define-constant ERR-MINT-PASS-LIMIT-REACHED (err u109))
+(define-constant ERR-ADD-MINT-PASS (err u110))
+
+
+(define-constant ERR-NOT-AUTHORIZED (err u401))
+(define-constant ERR-NOT-OWNER (err u402))
+(define-constant ERR-NOT-V1-OWNER (err u403))
+(define-constant ERR-NOT-ADMINISTRATOR (err u404))
+
+(define-non-fungible-token crashpunks-v2 uint)
+
+;; data structures
+
+;; {owner, operator, nftIndex} -> boolean
+;; if {owner, operator, nftIndex} is in map, then operator can perform actions on behalf of owner for this nftIndex
+(define-map approvals {owner: principal, operator: principal, nft-index: uint} bool)
+(define-map approvals-all {owner: principal, operator: principal} bool)
+
+;; nftIndex -> price (in ustx)
+;; if nftIndex is not in map, it is not listed for sale
+(define-map nft-market uint {price: uint})
+
+;; whitelist address -> # they can mint
+(define-map mint-pass principal uint)
+
+;; SIP-09: get last token id
+(define-read-only (get-last-token-id)
+  (ok (- (var-get mint-counter) u1))
+)
+
+;; SIP-09: URI for metadata associated with the token
+(define-read-only (get-token-uri (id uint))
+    (ok (some (var-get base-uri)))
+)
+
+;; SIP-09: Gets the owner of the 'Specified token ID.
+(define-read-only (get-owner (nftIndex uint))
+  (ok (nft-get-owner? crashpunks-v2 nftIndex))
+)
+
+;; SIP-09: Transfer
+(define-public (transfer (nftIndex uint) (owner principal) (recipient principal))
+    (begin
+        (asserts! (unwrap! (is-approved nftIndex contract-caller) ERR-NOT-AUTHORIZED) ERR-NOT-AUTHORIZED)
+        (asserts! (is-none (map-get? nft-market nftIndex)) ERR-NFT-LISTED)
+        (nft-transfer? crashpunks-v2 nftIndex owner recipient)
+    )
+)
+
+;; operable
+(define-read-only (is-approved (nftIndex uint) (operator principal))
+    (let ((owner (unwrap! (nft-get-owner? crashpunks-v2 nftIndex) ERR-COULDNT-GET-NFT-OWNER)))
+        (ok (is-owned-or-approved nftIndex operator owner))
+    )
+)
+
+;; operable
+(define-public (set-approved (nftIndex uint) (operator principal) (approved bool))
+    (let ((owner (unwrap! (nft-get-owner? crashpunks-v2 nftIndex) ERR-COULDNT-GET-NFT-OWNER)))
+        (ok (map-set approvals {owner: contract-caller, operator: operator, nft-index: nftIndex} approved))
+    )
+)
+
+(define-public (set-approved-all (operator principal) (approved bool))
+    (ok (map-set approvals-all {owner: contract-caller, operator: operator} approved))
+)
+
+;; public methods
+
+;; upgrade from v1 to v2
+;; Owner of crashpunks v1 calls this upgrade function
+;; 1. Transfers v1 NFT to this contract
+;; 2. This contract mints the v2 NFT with the same nftIndex for contract-caller
+;; 3. This contract burns the v1 NFT
+(define-public (upgrade-v1-to-v2 (nftIndex uint))
```

Comment:
What happens if v1 is minted again after an upgrade?

Reply:
v1 minting is essentially shut off anyway, but are you asking if it's possible to upgrade nft id 0, and then remint nft id 0 for v1?